### PR TITLE
grt/odb: check if net is connected by abutment

### PR DESCRIPTION
**Pull request by eder-matheus**

### CODE REVIEW

File: src/grt/.gitignore [hunk 1]
```diff
@@ -11,5 +11,6 @@ test/src/*/out.guide
 test/src/*/out2.guide
 test/src/*/route.guide
 test/results
+test/failures_report.json
```
Comment (maliberty) [inline src/grt/.gitignore:14]: What creates this?  I'm wondering if other modules would be the same and we should put it at the OR level.

File: src/odb/src/db/dbNet.cpp [hunk 2]
```diff
@@ -1801,6 +1801,50 @@ void dbNet::clearSpecial()
...
+  int pin_cnt = 0;
+  std::vector<Rect> first_pin_boxes;
+  for (dbITerm* iterm : getITerms()) {
+    dbMTerm* mterm = iterm->getMTerm();
+    dbMaster* master = mterm->getMaster();
+    if (!master->isBlock()) {
+      return false;
+    }
+
+    dbInst* inst = iterm->getInst();
```
Comment (maliberty) [inline src/odb/src/db/dbNet.cpp:1819]: Check that the instance is placed.

File: src/odb/src/db/dbNet.cpp [hunk 3]
```diff
@@ -1801,6 +1801,50 @@ void dbNet::clearSpecial()
   }
 }
 
+bool dbNet::isConnectedByAbutment()
+{
+  if (getITermCount() > 2 || getBTermCount() > 0) {
+    return false;
+  }
+
+  int pin_cnt = 0;
```
Comment (maliberty) [inline src/odb/src/db/dbNet.cpp:1810]: "bool first_mterm = true" seems more descriptive.

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/grt/src/GlobalRouter.cpp [hunk 4]
```diff
@@ -841,6 +841,7 @@ void GlobalRouter::initNets(std::vector<Net*>& nets)
   fastroute_->setMaxNetDegree(max_degree);
 
   if (verbose_) {
+    min_degree = min_degree == std::numeric_limits<int>::max() ? 0 : 1;
```
[the remark this answers is not in the document]
Comment (eder-matheus) [reply src/grt/src/GlobalRouter.cpp]: It was a mistake. I just wanted to have the degrees as zero when there is no net to routing. Now it should be fixed.

File: src/odb/src/db/dbNet.cpp [hunk 5]
```diff
@@ -1817,29 +1817,31 @@ bool dbNet::isConnectedByAbutment()
...
-        }
-        odb::Rect rect = box->getBox();
-        transform.apply(rect);
-        if (pin_cnt == 0) {
-          first_pin_boxes.push_back(rect);
-        } else {
-          for (const Rect& first_pin_box : first_pin_boxes) {
-            if (rect.intersects(first_pin_box)) {
-              return true;
+    if (inst->isPlaced() || inst->isFixed()) {
```
Comment (maliberty) [inline src/odb/src/db/dbNet.cpp:1820]: isPlaced is a superset of isFixed so no need for the ||...

File: src/odb/src/db/dbNet.cpp [hunk 6]
```diff
@@ -1817,29 +1817,31 @@ bool dbNet::isConnectedByAbutment()
...
+            for (const Rect& first_pin_box : first_pin_boxes) {
+              if (rect.intersects(first_pin_box)) {
+                return true;
+              }
             }
           }
         }
       }
     }
-    pin_cnt++;
```
Comment (maliberty) [inline src/odb/src/db/dbNet.cpp]: first_mterm = false ?

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
@oharboe this should resolve your overflow issue

**Comment by oharboe** [comment]
> @oharboe this should resolve your overflow issue

This PR makes sense to me...

Remind me of the test case where I ran into this?

**Comment by maliberty** [comment]
You had a thin line of overflow markers in your array

**Comment by maliberty** [comment]
https://github.com/Pinata-Consulting/pii-ascenium/issues/57

**Comment by oharboe** [comment]
> You had a thin line of overflow markers in your array

Ah, yes, I recall. Changing the size of the macros slightly larger removed the error so I was able to test the rest of the flow.

Thanks!

**Pull request merged**